### PR TITLE
Added environment variable to specify remote maven agent address

### PR DESCRIPTION
**Pull request by dump247**
I am working on a plugin to run jobs in a docker container. This change is necessary to run maven builds in the container since loopback and the local hostname will refer to the container rather than the host system where the jenkins agent service is running.

### CODE REVIEW

**Comment by ndeloof** [comment]
@dump247 did you found any solution to this issue ? I have same need for docker-custom-build-environment

**Comment by dump247** [comment]
@ndeloof I went in a completely different direction with my plugin. The plugin now inserts itself as a load balancer and spins up a container running the slave jar when a job is launched. This has some advantages over implementing as part of the jenkins job launch process. Also, it is completely transparent to the jenkins jobs and you don't run into issues like this.

I need to work on documentation, but it's been very successful in our jenkins installation: https://github.com/dump247/jenkins-docker-build-plugin

**Comment by ndeloof** [comment]
@dump247 From a quick look into your source code, you connect to docker slave image using ssh so you can establish the slave channel. So this look very similar to  [docker-plugin](https://github.com/jenkinsci/docker-plugin).  
How does this compare ?

**Comment by dump247** [comment]
@ndeloof  It does not log into the container itself. It uses ssh to log into the host machine to access the docker api. It also uses ssh to transfer the slave jar and some scripts to use when launching containers. I initially went with accessing the docker api remotely via ssl, but there were a couple of issues:

# There appears to be an issue with golang and ssl client keys. Eventually, the connection stops sending/receiving on the go side. I don't know what was going on, but it was easy to replicate with a standalone go application.
# I was using the docker log api to communicate with the slave jar running in the container, but docker retains that output. I wanted to reuse the container when running the job again, so the result was the slave jar output was piling up on the host system.

Basically, the plugin logs into the host machine and launches the slave jar in a docker container. The container is for a specific job and is shut down once the job completes. The slave jar is mapped into the container with a mapped directory. The slave jar has an option to specify the IP/port to connect to, so I use that to connect back to the script that launched the container and forward that communication to the master. The result is that the job can use whatever docker image it wants so long as it has a JRE installed.

**Comment by ndeloof** [comment]
ah ok, this is similar to another plugin I've been experimenting, which only do require docker image to include java runtime, and do run slave agent in docker container using downloadable remoting.jar.
But I have used docker API directly from Cloud API, not ssh to access slave nodes.

**Comment by dump247** [comment]
@ndeloof There are a couple of problems with using the remoting jar. 1. you need credentials and 2. it's tougher to map the slave node to the job that you launched the slave node for. Also, go's issue with client ssl certificates makes it impossible to secure the go api directly. With my plugin's SSH setup, you don't need jenkins credentials and the slave will only run the job it was launched for.

**Comment by ndeloof** [comment]
I'm passing JNLP lancher secret token as part of the slave launch command
I'm using a UID label to force the job to match this slave
I don't understand you go ssh issue, Docker API is TLS encrypted so does not require any extra security

**Comment by dump247** [comment]
I didn't know how to make use of the JNLP launcher like that. There is some small advantage in not having to expose that secret and not having to match UID to job, but JNLP approach has some things to like.

TLS provides the encryption, but not the access control. You can't prevent anyone from calling the docker APIs unless you incorporate client certificates. Although it's not terrible, it is a bit more of a hassle to setup than SSH. Also, like I said, go has some sort of bug when using client certificates where the TCP connection stops sending and receiving.

**Comment by oleg-nenashev** [comment]
I feel this pull request is no longer relevant...

* There are plenty of Docker plugins, including Docker Slaves plugin, which supposedly does what is expected to happen in this PR (@ndeloof please correct me if I am wrong).
* JNLP4 provides TLS encryption OOTB
* Docker offers much API security options
* Merge conflicts

@dump247 WDYT?

**Comment by ndeloof** [comment]
@oleg-nenashev IIUC this is comparable to https://github.com/jenkinsci/maven-interceptors/commit/a40fbf9203da36b00ebf9bbe93df561da4058807 and as such obsolete
